### PR TITLE
Improvements to how dvsim.py runs jobs

### PR DESCRIPTION
**Pull request by rswarbrick**
This patch queue tries to unpick some of the logic in `Deploy.py` which has got a little tangled. There are quite a few commits (9), because I wanted to merge incremental, bisectable changes rather than a giant "rewrite the world" commit.

The main thrust of the set of patches is to move the scheduling logic into a new file (`Runner.py`) and untangle all the "scheduley" bits from the code in the `Deploy` class. There are a few main changes:

1. Job tracking is now done in a scheduler class, rather than in the `Deploy` class, which also represents jobs to be run.
1. Jobs now have a notion of dependencies, rather than the ability to spawn new jobs. This matches other build systems much more closely, and also simplifies how we do coverage runs.
1. Signal handling (intercepting SIGINT) is now done by the scheduler rather than `dvsim.py`.

There are a few user-visible changes (and hopefully no bugs!):

1. If you have 100 tests that depend on 1 build and the build fails, you don't have to wait `100 / parallelism` seconds for dvsim to crank through the tests it can't run.
1. The timer doesn't reset to zero when starting coverage collection.

This refactoring also makes it possible to make other improvements in the future. For example, it would now be much simpler to add LSF job array support: we could essentially just dump the commands for a given target into shell scripts and then fire them all off at once. In the code, this would be represented by an `LSFTargetRunner` class (subclassing `TargetRunner`) with a specialized (and simplified!) `run` method.

Another possible future improvement (which is slightly at odds with LSF magic) is to allow different "targets" (the phases of the run) to run at once, which would allow you to fire off some runs as soon as their build was ready. This would make no difference on my laptop (with few cores and fewer EDA licences), but might make a significant difference on a big compute cluster, where you can get started on the big list of tests as soon as some can be run. Of course, this is getting to look awfully like a Make replacement: if we move further in that direction, I suggest dumping commands into shell scripts and writing out the dependency DAG to a Makefile.

@cindychip, @msfschaffner: As usual, I can't try out some of the linter commands or the FPV commands locally. If you have a minute, I'd be really grateful if you could try out this PR to make sure I haven't broken anything with a silly typo. I think that most of the cleverness is on the simulation side (which I could test), so I'm hopeful this will work.

### CODE REVIEW

**Comment by rswarbrick** [comment]
Force-push rebases on top of the code in PR #3948, which clashes with this PR. It also fixes a silly bug (I forgot to actually kill the item as well as moving it from the running to the killed list).

**Comment by rswarbrick** [comment]
Force-push fixes `--run-only` (before the fix, I'd accidentally left dependencies from the run jobs on the un-queued build jobs).

**Comment by rswarbrick** [comment]
Force-push rebases onto master (no longer depending on a merged PR)

**Comment by rswarbrick** [comment]
Force push has no interesting changes: just a weekly rebase to avoid file conflicts.

**Comment by cindychip** [comment]
Hey Rupert, I am super sorry for the late reply.. i tried it on both FPV and VeribleLint, I think they are working fine. And the changes look very cool to me. I think would be good to wait until Sri has a chance to review this.

**Comment by rswarbrick** [comment]
Hi Cindy, Thanks for the response. I think that Sri wants to take a very close look at any dvsim changes (we don't have any test suite and he understandably doesn't want to break this project-critical infrastructure). So I'm just rebasing once a week and leaving things in a working state until he has a chance to look at it.

**Comment by rswarbrick** [comment]
Force-push is just a weekly rebase to keep everything from bit-rotting.

**Comment by rswarbrick** [comment]
Another week, another rebase. No intentional changes, although I did have to spend some time fixing things up when rebasing past recent dvsim changes. It's a bit of a pity that this is gently bit-rotting.

File: util/dvsim/Runner.py
Comment (sriyerg) [inline util/dvsim/Runner.py:322]: Previously, `Deploy.deploy()` was invoked multiple times (later for coverage). Hence the `print_legend` flag was added. With these updates, it is no longer needed.

File: util/dvsim/Runner.py
Comment (sriyerg) [inline util/dvsim/Runner.py:72]: This probably defeats the purpose of ordering tests by iteration # in `SimCfg._expand_run_list`.

File: util/dvsim/SimCfg.py
Comment (sriyerg) [inline util/dvsim/SimCfg.py:530]: Also, if `self.runs` is empty, then there is nothing to run coverage on (try running `-i bogus_test --cov`).

File: util/dvsim/SimCfg.py
Comment (sriyerg) [inline util/dvsim/SimCfg.py]: Need to be able to handle the case where items is None (in case you end up mistakenly running a non-existent test).

**Review by sriyerg — COMMENTED** [review]
This PR is hard to review these massive changes. I am probably about 40% done. In subsequent commits, the diff is with your own code from previous commits in this PR! This work could be easily split into 3 distinct PRs (Timer, Runner and Results) containing logically separate enhancements to different parts of the tool. I'd appreciate if you could do that. If you'd rather not, then please allow some more time for me to go through it.

**Review by rswarbrick — COMMENTED** [review]
Hi Sri. Thanks for the review comments. I've addressed them (I think) in the latest force-push. I don't think that the PR splits very easily, or at least I think smaller PRs would end up clashing lots, so we'd still have to merge them in order.

Would you prefer me to split out an initial chunk, so that you can review it separately there? The first 2 commits are kind of trivial: maybe they'd be a good initial candidate?

File: util/dvsim/Deploy.py
Comment (rswarbrick) [reply util/dvsim/Deploy.py]: Good point (and it avoids some churn in later commits too). Done, thanks.

File: util/dvsim/Runner.py
Comment (rswarbrick) [reply util/dvsim/Runner.py]: I agree. I've called it `Scheduler` (and I *think* I've expunged all occurrences of the previous name!)

File: util/dvsim/Deploy.py
Comment (rswarbrick) [reply util/dvsim/Deploy.py]: Oops, yes! Thanks!

File: util/dvsim/Runner.py
Comment (rswarbrick) [reply util/dvsim/Runner.py]: Oh, good point. I've changed `_queued` to be a list, which should ensure everything gets popped in order.

File: util/dvsim/SimCfg.py
Comment (rswarbrick) [reply util/dvsim/SimCfg.py]: You're right. I've changed the test to `... and self.runs`, which should do what we want.

File: util/dvsim/SimCfg.py
Comment (rswarbrick) [reply util/dvsim/SimCfg.py]: I don't think that `items` can ever be None. Looking at the code, it's always a list (`self.deploy` in `SimCfg._gen_results`, which is in turn set to a list in `SimCfg._create_deploy_objects`).

File: util/dvsim/dvsim.py
Comment (rswarbrick) [reply util/dvsim/dvsim.py:581]: I imagined that a `FlowCfg` was the "configuration for a flow" (which doesn't include the results of running it). If you believe it would be cleaner to store those results on the object, I'm happy to do that instead. What do you think?

**Comment by rswarbrick** [comment]
Force-push re-does the `_queued` change (it only needed to change to a list once we split up the scheduler).

**Comment by sriyerg** [comment]
> Another week, another rebase. No intentional changes, although I did have to spend some time fixing things up when rebasing past recent dvsim changes. It's a bit of a pity that this is gently bit-rotting.

It indeed is unfortunate. It has been already communicated that non-urgent / non-bug-fix enhancements like these, especially when its a mishmash of commits that make changes to several different parts of the code, will be reviewed with a much lower priority. There are at least 4 discrete meaningful changes in here that could have been submitted as separate PRs, which would have been a lot easier / smoother to review. Separate commits in the same PR is not at all helpful, if the diff in latter commits is against your own code from previous commits! 

I have already communicated before - large, non-urgent PRs will be looked at with a lower priority (expect an SLO of ~60 days). If that is unacceptable, please justify. If it is urgent, please file an issue so that the priority is understood. 

@domrizz0 @asb @sjgitty @tomroberts-lowrisc

**Comment by rswarbrick** [comment]
@sriyerg: Would you prefer me just to close this PR? I suggested splitting it up yesterday, but you've replied to a previous message. What would you like me to do? Obviously, I don't want to cause more work for you (or unnecessary friction). And I'm sorry for letting any frustration appear in a GitHub message.

**Comment by sriyerg** [comment]
> @sriyerg: Would you prefer me just to close this PR? I suggested splitting it up yesterday, but you've replied to a previous message. What would you like me to do? Obviously, I don't want to cause more work for you (or unnecessary friction). And I'm sorry for letting any frustration appear in a GitHub message.

No need to close this PR (or the others) - I am reviewing this one now. I think these are pretty good enhancements. To be clear, my intention is certainly not to discourage anyone from making improvements to the tool. It's just that it will take some time for me to get to them. How long depends on how easy it is to review the changes. 

> Hi Sri. Thanks for the review comments. I've addressed them (I think) in the latest force-push. I don't think that the PR splits very easily, or at least I think smaller PRs would end up clashing lots, so we'd still have to merge them in order.
> 
> Would you prefer me to split out an initial chunk, so that you can review it separately there? The first 2 commits are kind of trivial: maybe they'd be a good initial candidate?

Like you said, those are trivial - I am fine either way. I am done reviewing those 2. I am looking at the latter commits now.

**Comment by rswarbrick** [comment]
Force-push is just a rebase over some recent dvsim changes that `git` couldn't resolve automatically. No functional change.

**Comment by rswarbrick** [comment]
Force-push rebases over Weicai's `cov_unr` changes.

**Comment by rswarbrick** [comment]
Force-push is just a weekly rebase to keep everything applying cleanly.

**Comment by rswarbrick** [comment]
Force-push rebases over the change to how we load configs.

**Comment by rswarbrick** [comment]
Force-push rebases over recent remote scratch changes.

**Comment by rswarbrick** [comment]
Happy New Year! Force-push is just a rebase.

**Comment by sriyerg** [comment]
LGTM, will take the fix-if-something broke approach (nothing seems to be broken for now). Thanks for these changes. The splitting of the scheduling logic was definitely needed, which will be a step towards supporting a wide variety of load sharing tools in a more natural way. 

At Google, LSF is getting deprecated. We need to completely switch to using GCP within a month's timeframe. The way jobs interact with GCP is of proprietary nature and hence, cannot be exposed to the open repo. However, it is not too different from how LSF works, so it should not be too hard. 

The way I am envisioning it - we need to support local dispatch and LSF as backends, both of which can live in the open source repo. In addition, if a company wants to use their own compute dispatch method (GCP in our case), they should be able to easily extend the `TargetScheduler` class and override (or add) appropriate APIs to be able to interact with the dispatched job. This extended class will be private to the company. The goal is to make it easy to get dvsim to dispatch to AWS, Azure etc. in future based on each OT partner's needs. I will be making updates to that logic to this end in the coming weeks. 

I suppose you have some more ideas on improving dvsim. It will be great if we could document what's being planned and when is it expected to land, so that we are not blindsided by each other's updates. Let me know how do you want to track it. Thanks!

**Review by sriyerg — APPROVED** [review]

**Comment by rswarbrick** [comment]
Thanks for the review, Sri. About migrating from LSF: I'd really suggest doing something a bit like an LSF job array. So the scheduler just dispatches a list of jobs and waits until they've all finished. This can be piped into something like `xargs` to run locally, LSF if you have it (we don't), or presumably whatever GCP script you're going to use at Google. I think this is probably a nice simple interface that won't need too much magic glue code.

**Pull request merged**
